### ISSUE TITLE
[Feature Request] Add Podman container runtime support

### ISSUE THREAD

**Issue by m7medVision** [Dokploy/dokploy#134]
I would like to request support for [Podman](https://podman.io/) as a container runtime option, alongside Docker. [Podman](https://podman.io/) offers a similar container management experience with a focus on rootless containers, enhancing security.

**Comment by Siumauricio** [comment]
Hello, thank you for your request, however I am unable to proceed with this request for multiple reasons:

1. This would require rewriting and adding another container handler that does the same job as docker, the advantages could be minimal.
2. It does not support docker swarm which is essential for multi-node container management, what I saw it supports is something with kubernetes.
3. This would cause multiple incompatibilities in the current configurations and deployments and would be overloading the codebase we have.



For the moment I will close it, if in the future we decide to migrate either for X or Y reason this issue will be reopened, thanks!

**Comment by Malix-Labs** [comment]
_@Siumauricio could you close that as not planned instead please ? :)_

**Comment by jalberto** [comment]
@Siumauricio maybe I am being jsut a bit picky here, but this text copied from the website is clashing with the use of docker and docker swarm (both products managed by private companies)

> No Vendor Lock-In
> 
> Experience complete freedom to modify, scale, and customize Dokploy to suit your specific needs.

also ins your 1st point there:

> This would require rewriting and adding another container handler that does the same job as docker, the advantages could be minimal.

the advantages are not minimal

**Comment by Malix-Labs** [comment]
docker engine and docker swarm is FOSS, only docker desktop is proprietary

Since you do not need to lock into a vendor, regardless of it is is owned by a private company, "No Vendor Lock-In" is an accurate statement

> the advantages are not minimal

I am a podman enjoyer, but you need to elaborate

**Comment by jalberto** [comment]
Docker licences are a bit convulsive, there is a clear separation between Moby and docker, and there are different docker editions, but I guess it depends on the point of view.

About the advantages, just the possibility of running rootless containers and all it implies from security PoV is a big advantage, then we can add on top the play/kubernetes features that open a nice scalability path, also podman-desktop is opensource.

**Comment by Malix-Labs** [comment]
> Docker licences are a bit convulsive

I do not get what you mean

> About the advantages, just the possibility of running rootless containers and all it implies from security PoV is a big advantage, then we can add on top the play/kubernetes features that open a nice scalability path

100% agree

However, it is also to be acknowledged that it needs a substantial refactor for Dokploy (as said earlier by @Siumauricio : https://github.com/Dokploy/dokploy/issues/134#issuecomment-2156135806)

> podman-desktop is opensource

True but even in the case that Dokploy refactor to make podman work, I don't know if podman desktop would be of any use

**Comment by tyeongkim** [comment]
We need Podman support. Podman is recommended by Red Hat, so Podman is more handy in RHEL.

**Comment by frc4533-lincoln** [comment]
Additionally crun (used by Podman by default) is quite a bit more performant than runc (used by Docker).
<https://github.com/containers/crun#performance>

**Comment by Malix-Labs** [comment]
@Siumauricio do you think this could be reconsidered?

_Also read https://github.com/Dokploy/dokploy/issues/134#issuecomment-2453412022_

**Comment by f1yn** [comment]
As someone who's familiar with podman, docker and TypeScript - there's some notable performance benefits, but **substantial security benefits** to adding on podman abstraction.

The thing to be aware of, obviously, is that swarm wouldn't be possible - so it would potentially make more sense to have services build via podman as its own build type (instead of having it as a source).

The primary compatibility issue is Swarm not being [OCI compatible](https://opencontainers.org/). Podman however, has gone to great lengths to retain docker compatibility - even to the point of being able to [configure docker compose files into using podman containers directly via the environment.](https://www.redhat.com/en/blog/podman-docker-compose)

I don't think Swarm is a deal-breaker, but trying to get a partially-proprietary solution (Swarm) to scale with a fully open-source one (Podman) has been intentionally engineered to be difficult.

If I were to integrate this, I'd come from this from two angles:

1. Adding the ability to set podman as a build type. From what I've seen, this would be fair easier than a top down setting that would require substantial re-arch of the existing deploy structure.

  > This wouldn't be as beneficial for scale, but for security sensitive containers - podman runs laps around docker.

2. Revisiting the link I shared above, podman devs have gone to great lengths to remain as OCI compliant as possible. It might also be worth adding documentation so dokploy users can be informed that podman can be configured to be usable with docker compose files.

Again, not full coverage yet - but it doesn't need to be. Just enough to gauge usage, and see if managing podman containers via Dokploy is something worth adding.

---

I currently maintain over a dozen high utilization containers solely on podman, but I use cockpit and manual systemd configurations to do it. Having interoperability with Dokploy would a huge onboarding thing for me, namely because I could thin out my hardware requirements away from VMs (as I won't need to mitigate as many risks as I do with docker).

If there's specific TS files that would be useful for adding in Podman as a build type, I more than likely would love to give this a try (assuming it's a wanted addition).

**Comment by microfauna** [comment]
Has anyone tried running dokploy with podman's "docker compatibility" mode enabled and seen what happens? It routes all docker commands to podman under the hood, but pretends to be docker.

**Comment by Kynson** [comment]
I decided to get my hands dirty and set up Dokploy with podman.
TL;DR: It works (kind of)
Disclaimer: I only tested the following features and the result below is a big "It works on my machine".
## Setup (**single node only**)
Environment:
podman: v5.4.2
OS: Ubuntu Desktop 25.10 (aarch64 on VMWare Fusion)

I asked AI to generate the following docker-compose file based on the installation script:
```yml
services:
  dokploy-postgres:
    image: docker.io/library/postgres:16
    container_name: dokploy-postgres
    networks:
      - dokploy-network
    environment:
      POSTGRES_USER: dokploy
      POSTGRES_DB: dokploy
      POSTGRES_PASSWORD: amukds4wi9001583845717ad2
    volumes:
      - dokploy-postgres:/var/lib/postgresql/data
    restart: always

  dokploy-redis:
    image: docker.io/library/redis:7
    container_name: dokploy-redis
    networks:
      - dokploy-network
    volumes:
      - dokploy-redis:/data:Z
    restart: always

  dokploy:
    image: docker.io/dokploy/dokploy:latest
    container_name: dokploy
    networks:
      - dokploy-network
    ports:
      - '3000:3000'
    environment:
      - ADVERTISE_ADDR=${ADVERTISE_ADDR}
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      # Map the rootless Podman socket from the user runtime directory
      - ${XDG_RUNTIME_DIR}/podman/podman.sock:/var/run/docker.sock
      - /etc/dokploy:/etc/dokploy
      - /etc/dokploy-home:/root/.docker
    restart: always

  dokploy-traefik:
    image: docker.io/library/traefik:v3.6.1
    container_name: dokploy-traefik
    restart: always
    networks:
      - dokploy-network
    ports:
      - '8080:8080/tcp'
      - '3001:80/tcp'
      - '3002:443/tcp'
      - '3002:443/udp'
    environment:
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - /etc/dokploy/traefik/traefik.yml:/etc/traefik/traefik.yml
      - /etc/dokploy/traefik/dynamic:/etc/dokploy/traefik/dynamic
      - ${XDG_RUNTIME_DIR}/podman/podman.sock:/var/run/docker.sock

networks:
  dokploy-network:
    name: dokploy-network
    driver: bridge

volumes:
  dokploy-postgres:
  dokploy-redis:
  dokploy-home:
```
Note: I mapped port 3001 and 3002 to traefik as I don't want to deal with privileged ports (and keep podman rootless) 

1. Install `podman` and `podman-docker`
2. Create directory `/etc/dokploy` and `/etc/dokploy-home`
3. Change owner of the above directory to the user running podman
4. Run `systemctl --user status podman.socket` to check if the podman socket is running, if not, start it
5. Run `podman compose up` on the directory with the above docker-compose.yml
6. Wait the service to start and you may see traefik complaining it can't find docker swarm
7. Stop the compose with `podman compose down`
8. Edit `/etc/dokploy/traefik/traefik.yml` and remove the swarm provider 
9. Restart the compose

If everything goes well you should be able to visit `localhost:3000` and sign in to dokploy.

## What works (Based on what I tested)
- Deployment with docker compose file (I tested n8n and ntfy templates only)
- Scheduled tasks
- Monitoring
- Request logs 
- Container logs
- Access terminal to deployed container

Notes:
- Domains works but it seems that we shouldn't include the port (`3001`) in the domain rule or else traefik will give 404, but when you access the app locally, the port is required.
- I only tested this locally, with traefik.me resolving to `127.0.0.1`, not sure will this works on real server

## What does not work
- Swarm status page
- Cluster

Note: Remote server might work if the startup script is modified, but I did not test it.

**Comment by ccarcaci** [comment]
@Kynson 

A very similar implementation using Makefile to avoid podman-compose. I just guided AI to generate it and briefly tested, it has the same limitations as the solution above. Good for local development.

```
# Makefile for rsvr local infrastructure
# Runs Dokploy on macOS via plain podman commands, mounting the Podman socket
# directly into containers — no podman-compose or docker compose dependency.
#
# Bind mount behaviour on macOS / Podman Machine
# -----------------------------------------------
# Podman on macOS runs containers inside a hidden Linux VM (Podman Machine).
# Bind mounts resolve paths inside that VM, not on the macOS host.
# Only a few host directories are automatically shared with the VM:
#   ~/  (home directory), /Users, /Volumes
# Paths under /etc are NOT shared, so bind-mounting /etc/dokploy* fails
# with "statfs … no such file or directory".
#
# All persistent directories that previously lived under /etc/dokploy* are
# replaced with named volumes, which Podman manages entirely inside the VM.
# The traefik.yml static config is seeded into a named volume using a
# short-lived Alpine container so the file never needs to touch the host.

# IP address Dokploy advertises to Traefik for routing.
# Override at the CLI if your machine has a non-loopback address:
#   make start ADVERTISE_ADDR=192.168.1.10
ADVERTISE_ADDR ?= 127.0.0.1

# Absolute path to this Makefile's directory — used to locate traefik.yml.
INFRA_DIR := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))

# Default target — show help
.DEFAULT_GOAL := help

.PHONY: help
help: ## Show this help message
	@echo "rsvr - Local Infrastructure (Dokploy via plain podman)"
	@echo ""
	@echo "Usage: make [target]"
	@echo ""
	@echo "Available targets:"
	@awk 'BEGIN {FS = ":.*##"; printf ""} /^[a-zA-Z_-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)

##@ Lifecycle

.PHONY: start
start: ## Start the Dokploy stack (Podman must be running)
	@PODMAN_SOCK=$$(podman info --format '{{.Host.RemoteSocket.Path}}' 2>/dev/null | sed 's|unix://||'); \
	if [ -z "$$PODMAN_SOCK" ]; then \
		echo "Could not detect Podman socket. Is Podman running?"; \
		exit 1; \
	fi; \
	\
	# ------------------------------------------------------------------ \
	# Named volumes                                                        \
	# All volumes are managed by Podman inside the Linux VM.              \
	# No host paths under /etc are used — those are not visible to the VM.\
	# ------------------------------------------------------------------ \
	podman volume exists dokploy-postgres 2>/dev/null || \
		podman volume create dokploy-postgres; \
	podman volume exists dokploy-redis 2>/dev/null || \
		podman volume create dokploy-redis; \
	# dokploy-home  →  /root/.docker inside the dokploy container \
	podman volume exists dokploy-home 2>/dev/null || \
		podman volume create dokploy-home; \
	# dokploy-config  →  /etc/dokploy inside the dokploy container \
	podman volume exists dokploy-config 2>/dev/null || \
		podman volume create dokploy-config; \
	# dokploy-traefik-config  →  /etc/traefik inside the traefik container \
	# Seed traefik.yml into the volume using a short-lived Alpine container. \
	# The -c flag passes a shell command string; cat writes the heredoc content. \
	podman volume exists dokploy-traefik-config 2>/dev/null || { \
		podman volume create dokploy-traefik-config; \
		podman run --rm \
			--volume dokploy-traefik-config:/etc/traefik \
			--volume "$(INFRA_DIR)traefik.yml:/tmp/traefik.yml:ro" \
			docker.io/library/alpine:3.21 \
			sh -c "cp /tmp/traefik.yml /etc/traefik/traefik.yml"; \
	}; \
	# dokploy-traefik-dynamic  →  /etc/dokploy/traefik/dynamic inside traefik \
	podman volume exists dokploy-traefik-dynamic 2>/dev/null || \
		podman volume create dokploy-traefik-dynamic; \
	\
	# ------------------------------------------------------------------ \
	# Network                                                              \
	# ------------------------------------------------------------------ \
	podman network exists dokploy-network 2>/dev/null || \
		podman network create --driver bridge dokploy-network; \
	\
	# ------------------------------------------------------------------ \
	# Containers                                                           \
	# Start existing containers; create and run them if they don't exist. \
	# ------------------------------------------------------------------ \
	if podman container exists dokploy-postgres 2>/dev/null; then \
		podman start dokploy-postgres; \
	else \
		podman run --detach \
			--restart=always \
			--name dokploy-postgres \
			--network dokploy-network \
			--env POSTGRES_USER=dokploy \
			--env POSTGRES_DB=dokploy \
			--env POSTGRES_PASSWORD=amukds4wi9001583845717ad2 \
			--volume dokploy-postgres:/var/lib/postgresql/data \
			docker.io/library/postgres:16; \
	fi; \
	if podman container exists dokploy-redis 2>/dev/null; then \
		podman start dokploy-redis; \
	else \
		podman run --detach \
			--restart=always \
			--name dokploy-redis \
			--network dokploy-network \
			--volume dokploy-redis:/data:Z \
			docker.io/library/redis:7; \
	fi; \
	if podman container exists dokploy 2>/dev/null; then \
		podman start dokploy; \
	else \
		podman run --detach \
			--restart=always \
			--name dokploy \
			--network dokploy-network \
			--publish 3000:3000 \
			--env ADVERTISE_ADDR=$(ADVERTISE_ADDR) \
			--env DOCKER_HOST=unix:///var/run/docker.sock \
			--volume "$$PODMAN_SOCK":/var/run/docker.sock \
			--volume dokploy-config:/etc/dokploy \
			--volume dokploy-home:/root/.docker \
			docker.io/dokploy/dokploy:latest; \
	fi; \
	if podman container exists dokploy-traefik 2>/dev/null; then \
		podman start dokploy-traefik; \
	else \
		podman run --detach \
			--restart=always \
			--name dokploy-traefik \
			--network dokploy-network \
			--publish 8080:8080/tcp \
			--publish 3001:80/tcp \
			--publish 3002:443/tcp \
			--publish 3002:443/udp \
			--env DOCKER_HOST=unix:///var/run/docker.sock \
			--volume dokploy-traefik-config:/etc/traefik \
			--volume dokploy-traefik-dynamic:/etc/dokploy/traefik/dynamic \
			--volume "$$PODMAN_SOCK":/var/run/docker.sock \
			docker.io/library/traefik:v3.6.1; \
	fi
	@echo "Dokploy starting at http://localhost:3000"

.PHONY: stop
stop: ## Stop the Dokploy stack
	@podman stop dokploy-traefik dokploy dokploy-redis dokploy-postgres

.PHONY: restart
restart: ## Restart the Dokploy stack
	@podman restart dokploy-traefik dokploy dokploy-redis dokploy-postgres

.PHONY: logs
logs: ## Tail logs from all containers
	@podman logs --follow dokploy-postgres & \
	podman logs --follow dokploy-redis & \
	podman logs --follow dokploy & \
	podman logs --follow dokploy-traefik & \
	wait

##@ Removal

.PHONY: delete
delete: ## Destroy the stack and permanently delete ALL Dokploy data
	@echo "WARNING: This will permanently remove all Dokploy containers and volumes."
	@echo "ALL data will be lost: databases, services, and configuration."
	@echo ""
	@printf "Type 'yes' to confirm: "; \
	read CONFIRM; \
	if [ "$$CONFIRM" != "yes" ]; then \
		echo "Aborted. Nothing was removed."; \
		exit 1; \
	fi
	@echo ""
	@podman stop dokploy-traefik dokploy dokploy-redis dokploy-postgres 2>/dev/null || true
	@podman rm dokploy-traefik dokploy dokploy-redis dokploy-postgres 2>/dev/null || true
	@podman volume rm \
		dokploy-postgres \
		dokploy-redis \
		dokploy-home \
		dokploy-config \
		dokploy-traefik-config \
		dokploy-traefik-dynamic \
		2>/dev/null || true
	@podman network rm dokploy-network 2>/dev/null || true
```
